### DISCUSSION TITLE
compose distort image gets fragmented

### DISCUSSION THREAD

**Discussion by MarliesMaalderink** [ImageMagick/ImageMagick#3213] [Help] [answered]
Hi everyone,

I hope someone can help me because I am lost...

I am trying to create bed sheets (duvet covers) in image magick. I have purchased a mockup file for photoshop, which works with 3D layers so that's not much use when using image magick, but I am basically using it just as an overlay.

I have placed my source image inside the mockup using this:

 magick source.png displace.jpg -virtual-pixel transparent -compose Distort -composite result.png

And overal it turned out exactely as I wanted too, except for that the image lost it's sharpness, all edges of the sea creatures on the images are sort of fragmented and I have no clue at all how to fix this.

I have been using this page for documentations:
https://legacy.imagemagick.org/Usage/mapping/
Somewhere in the middle is an example of a photo of a mandril that is made to look round, and in this example this image doens't seem so fragmented but it's hard to say if that is because it's such a small image...

Here's the endresult:
http://www.marliesmaalderink.nl/images/testRender.jpg (all sea creatures should have sharp, crisp edges)
and here's what I use as a displacement map:
http://www.marliesmaalderink.nl/images/testDisplace.jpg

Is there a way to distort an image so that the source image still keeps it's original sharpness rather then it getting all fragmented?

**Comment by snibgo** [comment]
What version of IM, on what platform? Please paste the text output of "magick -version".

Please show your input image "source.png".

Your displacement map testDisplace.jpg is JPEG. I recommend that you do not use JPG compression for any image processing, because it is only 8 bit/channel, and the compression is lossy. With 8 bits, the precision of displacement is 1 part in 256. Your testRender.jpg output is clearly much larger than 256 pixels wide, so you get a spatial quantization effect in the position of pixels, "sort of fragmented".

So the first fix is to use a displacement map that has at least 16 bits of precision. More is better, eg 32-bit floats.

If for some reason this is not possible, then perhaps supersampling, possibly with a small blur, may help.

**Reply by MarliesMaalderink** [reply]
Hi snigbo:

I am using:
Version: ImageMagick 7.0.10-29 Q16 x86_64 2020-09-10 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5) 
Delegates (built-in): freetype jng jpeg png x xml zlib

On ubuntu 20.04 (on this particular render)

The original image:
https://www.marliesmaalderink.nl/images/ocean.png (it's probably copyrighted by www.patterndesigns.com but it was the most representive image I had for now)

thank you so much for your recommendation, which make sense to me, so I will start experimenting with that first. I am making the displacement map in photoshop. This might be a silly question, but would this be the way to go?:
Create file in photoshop with Mode set to RGB color and 32 Bits/Channel, create the map, and then save as png?

**Comment by snibgo** [comment]
You "magick -version" says you have HDRI (ie IM stores pixels as floating point), so that is good.

For https://www.marliesmaalderink.nl/images/ocean.png, I get:
```Warning: Potential Security Risk Ahead

Firefox detected a potential security threat and did not continue to www.marliesmaalderink.nl. If you visit this site, attackers could try to steal information like your passwords, emails, or credit card details.

What can you do about it?

The issue is most likely with the website, and there is nothing you can do to resolve it. You can notify the website’s administrator about the problem.
```
Changing "https://" to "http://" fixes that problem.


I don't use Photoshop. PNG is limited to 16 bits/channel, and integer only. If Photoshop can save TIFF as 32-bit floating point, that would be better.

Blurring your JPG map does help, eg:
```
magick ocean.png ( testDisplace.jpg -blur 0x2 ) -virtual-pixel transparent -compose Distort -composite r.png
```
... but starting with a more precise map would be better.

**Reply by MarliesMaalderink** [reply]
So sorry about the ssh error! It should have been http... 

Anyway,  thank you so much for helping!

So with the help of your tips I managed to get the resulting image so much better:
http://www.marliesmaalderink.nl/images/testRender2.png

I am still using a png as displacement here - I made a 32 bit tiff but then found that my IM installation didn't support tif so I suppose I have to reinstall IM first with the tif delegate. I hope this will fix the pixels near the edges.

**Comment by snibgo** [comment]
> Delegates (built-in): freetype jng jpeg png x xml zlib

Your list of delegates is rather small. Here is the list for IM v7.0.10-45, the pre-built IM for Windows:
```
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr
 lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib
```
You might find jp2, lcms, openexr, and raw useful.

**Reply by MarliesMaalderink** [reply]
Thanks! In the meanwhile I reinstalled  imagemagick with required delegates. And I noticed that when using a 32 bit tiff, the pixelation is almost entirely gone, so that's good. But it has been distorted quite differently from the png, even though the distortion maps look the same to the naked eye (yet their internally different I guess since one is a 32 bit tiff and the other a 8 bit png)

This is the new distorted file:
http://www.marliesmaalderink.nl/images/distortionWithtif.png
The old one:
http://www.marliesmaalderink.nl/images/testRender2.png

I guess I can trial and error my way out of here from now but I do have one more question if I may:
You say you don't use photoshop. How would you create a displacement map like this then? I found it very inconvenient in photoshop because once I set the file to 32 bit, I can't edit it anymore. So I have to create the image and then change it to 32 bit just for saving it, but I'm not sure what this does to the colors. (this actually seems to be a bug in photoshop that has not been solved for the past 8 years)

**Comment by snibgo** [comment]
If the maps look the same to the naked eye, then the results should look roughly the same. (A small spatial displacement is more obvious than a small colour shift, which is why we need high-precision displacement maps.) Your results are very different.

There may be a compatibility problem in the 32-bits files. Can you link to the 32-bit map? Both exiftool and "identify -verbose" should say it is 32-bit floating point.

I usually make displacement maps as part of an automated process. I make them manually in Gimp. Gimp and IM can exchange 32-bit floating-point TIFFs with no problem.

**Reply by MarliesMaalderink** [reply]
Hmmm, odd... 

this is the tif:
http://www.marliesmaalderink.nl/images/displace.tif

and the png:
http://www.marliesmaalderink.nl/images/displace.png

**Reply by snibgo** [reply]
Exiftool and IM both say displace.tif is 16-bit integer. They also say displace.png is only 8-bit integer.

**Reply by MarliesMaalderink** [reply]
Hmmm, then I guess photoshop is doing something unexpected, since I exported a 32 bit psd file to a tif using 32 bit settings... I will have to find out why that is...

**Comment by snibgo** [comment]
If the maps look the same to the naked eye, then the results should look roughly the same. (A small spatial displacement is more obvious than a small colour shift, which is why we need high-precision displacement maps.) Your results are very different.

There may be a compatibility problem in the 32-bits files. Can you link to the 32-bit map? Both exiftool and "identify -verbose" should say it is 32-bit floating point.

I usually make displacement maps as part of an automated process, but sometimes manually with Gimp. Gimp and IM can exchange 32-bit floating-point TIFFs with no problem.

**Reply by fmw42** [reply]
FYI, I have had problems with exchanging PS 32 TIFF with ImageMagick. But I am on an older PS CS version. So I thought perhaps it was the older version.

**Reply by MarliesMaalderink** [reply]
I am on photoshop CS6 and I seem to be experiencing a lot of problems... appearantly my 32 tiff is only 16 bit (while using a 32 bit  mode psd file and exporting it with the right settings) and I can't edit anything. I found this thread which seems to indicate that we're dealing with a 7 year old problem here...

https://feedback.photoshop.com/conversations/photoshop/photoshop-cc-unable-to-edit-32-bitchannel-images/5f5f45654b561a3d42487caa?page=1

I guess I'll try gimp then...

**Comment by snibgo** [comment, marked as answer]
As an example, using Gimp for a 32-bit floating-point map:

1. Create an identity displacement map. (Windows BAT syntax.)
```
magick ^
  -size 1000x1000 xc: ^
  -sparse-color Bilinear ^
0,0,#000,^
%%[fx:w-1],0,#f00,^
0,%%[fx:h-1],#0f0,^
%%[fx:w-1],%%[fx:h-1],#ff0 ^
  -define quantum:format=floating-point -depth 32 ^
  map.tiff
```
2. Verify the file has 4 channels (RGBA) of 32-bit floats:
```
f:\web\im>exiftool map.tiff

ExifTool Version Number         : 12.13
File Name                       : map.tiff
Directory                       : .
File Size                       : 11 MiB
File Modification Date/Time     : 2021:02:10 09:59:54+00:00
File Access Date/Time           : 2021:02:10 09:59:23+00:00
File Creation Date/Time         : 2021:02:10 09:59:23+00:00
File Permissions                : rw-rw-rw-
File Type                       : TIFF
File Type Extension             : tif
MIME Type                       : image/tiff
Exif Byte Order                 : Big-endian (Motorola, MM)
Image Width                     : 1000
Image Height                    : 1000
Bits Per Sample                 : 32 32 32
Compression                     : Uncompressed
Photometric Interpretation      : RGB
Fill Order                      : Normal
Strip Offsets                   : 8
Orientation                     : Horizontal (normal)
Samples Per Pixel               : 3
Rows Per Strip                  : 1000
Strip Byte Counts               : 12000000
Planar Configuration            : Chunky
Page Number                     : 0 1
White Point                     : 0.3127000032 0.328999996
Primary Chromaticities          : 0.6399999856 0.330000013 0.3000000118 0.600000
0237 0.1500000058 0.05999999844
Sample Format                   : Float; Float; Float
S Min Sample Value              : 0 0 0
S Max Sample Value              : 1 1 1
Image Size                      : 1000x1000
Megapixels                      : 1.0
```

3. Edit map.tiff in Gimp, v2.10.0, with various transform tools. Save it as map2.tiff. I have uploaded map2.tiff to http://snibgo.com/imforums/map2.tiff if you want to play with it.

4. Verify map2.tiff has 4 channels (RGBA) of 32-bit floats, using exiftool as above.

5. Use map2.tiff to distort the image:
```
magick ocean.png map2.tiff -virtual-pixel transparent -compose Distort -composite map2_r.png
```
map2_r.png is:
![map2_r](https://user-images.githubusercontent.com/33812368/107504425-2df00000-6b93-11eb-8233-9d8d39df2b89.png)
The result has no fragmentation.

**Reply by MarliesMaalderink** [reply]
Oh, this reply is so extremely helpful!! I just made the displacement file with the code you provided, and honestly, it looks rather different from my initial displacement file which I created in photoshop manually, so that might already have been part of the problem.

I finish up tomorrow (unfortunately I don't have time to finish it this afternoon) and let you know, I think with all the information you provided I will be able to fix this, but I'll let you know. Thank you so much for all your advice and help!!! I really appreciate it!

**Reply by MarliesMaalderink** [reply]
In addition to the above, I was looking at your exiftool output, but what do I have to look for here when determining whether it has 4 channels (RGBA) of 32-bit floats? (I was expecting something like "channels" and/or "Bit Depth" or something... )

**Comment by snibgo** [comment]
My previous post has a minor error: map.tiff has no alpha channel, but map2.tiff does.

From "exiftool map2.tiff", look for these:
```
Bits Per Sample                 : 32 32 32 32
Sample Format                   : Float; Float; Float; Float
```

**Reply by MarliesMaalderink** [reply]
So I was really hoping that it could have fixed it, but my problem seems to be persistent... 

First of all, when I use your displacementMap I get the same result as you do, it's sharp and crisp. But then this:

I created a displacement file with this bat script: (I only made it 1500x1500, didn't change anything else):
```
convert ^
  -size 1500x1500 xc: ^
  -sparse-color Bilinear ^
0,0,#000,^
%%[fx:w-1],0,#f00,^
0,%%[fx:h-1],#0f0,^
%%[fx:w-1],%%[fx:h-1],#ff0 ^
  -define quantum:format=floating-point -depth 32 ^
  map.tiff
```
  That got me this file:
http://www.marliesmaalderink.nl/images/map.tiff

I then opened it in Gimp, which I just installed, version 2.10.22

I transform it with the warp transform pencil tool, and save it (by choosing "export"  in the menu) with these settings: (sorry for the Dutch but I guess it's the same checkboxes in any language, just with a different name, It's just the default settings)
http://www.marliesmaalderink.nl/images/gimpDialog.jpg

which gets me this:
http://www.marliesmaalderink.nl/images/gimpDisplace3.tiff

Exiftools says:
```
> $ exiftool gimpDisplace3.tiff                                                                                                                                                  
ExifTool Version Number         : 11.88
File Name                       : gimpDisplace3.tiff
Directory                       : .
File Size                       : 26 MB
File Modification Date/Time     : 2021:02:11 10:14:24+01:00
File Access Date/Time           : 2021:02:11 10:14:42+01:00
File Inode Change Date/Time     : 2021:02:11 10:14:38+01:00
File Permissions                : rw-rw-r--
File Type                       : TIFF
File Type Extension             : tif
MIME Type                       : image/tiff
Exif Byte Order                 : Little-endian (Intel, II)
Image Width                     : 1500
Image Height                    : 1500
Bits Per Sample                 : 32 32 32
Compression                     : Uncompressed
Photometric Interpretation      : RGB
Orientation                     : Horizontal (normal)
Samples Per Pixel               : 3
Rows Per Strip                  : 128
X Resolution                    : 72
Y Resolution                    : 72
Planar Configuration            : Chunky
Page Name                       : map.tiff
Resolution Unit                 : inches
Software                        : GIMP 2.10.22
Modify Date                     : 2021:02:11 10:11:12
Subfile Type                    : Reduced-resolution image
Strip Offsets                   : 27005366
Strip Byte Counts               : 196608
Sample Format                   : Float; Float; Float
XMP Toolkit                     : XMP Core 4.4.0-Exiv2
Document ID                     : gimp:docid:gimp:e1ed4d77-8c6c-412b-8df2-a14123a151af
Instance ID                     : xmp.iid:d9ea3fa6-3bf0-49f3-a809-c0bbb721d8ca
Original Document ID            : xmp.did:0beeb74e-4d1d-4069-adfc-4b6be4d2f0ea
Api                             : 2.0
Platform                        : Windows
Time Stamp                      : 1613034685502188
Version                         : 2.10.22
Format                          : image/tiff
Date/Time Modified              : 2021:02:11 10:11:12
Creator Tool                    : GIMP 2.10
Location Created                : 
Location Shown                  : 
Artwork Or Object               : 
Registry ID                     : 
History Action                  : saved
History Changed                 : /
History Instance ID             : xmp.iid:90baf59a-876a-4d8a-aad0-6c458cdfbada
History Software Agent          : Gimp 2.10 (Windows)
History When                    : 2021:02:11 10:11:25
Image Supplier                  : 
Image Creator                   : 
Copyright Owner                 : 
Licensor                        : 
Color Space                     : sRGB
Profile CMM Type                : Little CMS
Profile Version                 : 4.3.0
Profile Class                   : Display Device Profile
Color Space Data                : RGB
Profile Connection Space        : XYZ
Profile Date Time               : 2021:02:11 08:49:08
Profile File Signature          : acsp
Primary Platform                : Microsoft Corporation
CMM Flags                       : Not Embedded, Independent
Device Manufacturer             : 
Device Model                    : 
Device Attributes               : Reflective, Glossy, Positive, Color
Rendering Intent                : Perceptual
Connection Space Illuminant     : 0.9642 1 0.82491
Profile Creator                 : Little CMS
Profile ID                      : 0
Profile Description             : GIMP built-in sRGB
Profile Copyright               : Public Domain
Media White Point               : 0.9642 1 0.82491
Chromatic Adaptation            : 1.04788 0.02292 -0.05022 0.02959 0.99048 -0.01707 -0.00925 0.01508 0.75168
Red Matrix Column               : 0.43604 0.22249 0.01392
Blue Matrix Column              : 0.14305 0.06061 0.71393
Green Matrix Column             : 0.38512 0.7169 0.09706
Red Tone Reproduction Curve     : (Binary data 32 bytes, use -b option to extract)
Green Tone Reproduction Curve   : (Binary data 32 bytes, use -b option to extract)
Blue Tone Reproduction Curve    : (Binary data 32 bytes, use -b option to extract)
Chromaticity Channels           : 3
Chromaticity Colorant           : Unknown (0)
Chromaticity Channel 1          : 0.64 0.33002
Chromaticity Channel 2          : 0.3 0.60001
Chromaticity Channel 3          : 0.15001 0.06
Device Mfg Desc                 : GIMP
Device Model Desc               : sRGB
Image Size                      : 1500x1500
Megapixels                      : 2.2
Thumbnail TIFF                  : (Binary data 196824 bytes, use -b option to extract)
```

In which the Bits Per Sample and the Sample Format seem ok

then I dothis:
```magick ocean.png -resize 1500x1500! gimpDisplace3.tiff -compose Distort -composite testFile3.png```

makes me end up with this:
http://www.marliesmaalderink.nl/images/testFile3.png

(I realized after uploading and typing everything that I should make sure it's not because of the resize, so I checked this by doing everyting again in 1000x1000, but it doesn't make a difference)

Is there anything you can see about my files that might be wrong?

**Comment by snibgo** [comment]
The problem is in map.tiff. If you use that as the displacement map, it should not change the image. But it does.

map.tiff is 32-bit floating point, but the data is precise to only 8 bits, and it is integer.

You create map.tiff with "convert" instead of "magick". Why? What does "convert -version" say? Perhaps that version of IM is Q8 integer (instead of HDRI).

**Reply by MarliesMaalderink** [reply]
Oooh god... I never realized that. I created that file yesterday on my windows machine with an old IM version... It won't work if I convert it to a shell script and run it on Linux because "magick: invalid argument for option 'sparse-color': Invalid number of Arguments @ error/operation.c/SparseColorOption/252."

Is there a way to run this directly on the command line?

Guess I have to update my windows machine if it doesn't run on Linux... (I usually don't use IM on the windows machine)

**Reply by MarliesMaalderink** [reply]
never mind, already updated windows machine. It's ok now. I feel so stupid!! I totally forgot about the IM version issue when I typed that post.

Thank you so much for all your help and patience!!! I finally have a proper duvet cover now!

**Comment by snibgo** [comment]
In bash, percent % is not a special character, but hash # is. So don't escape the percent sign with another percent sign, but do escape the hash with a backslash: \\#.
```
magick.exe \
  -size 1000x1000 xc: \
  -sparse-color Bilinear \
0,0,\#000,\
%[fx:w-1],0,\#f00,\
0,%[fx:h-1],\#0f0,\
%[fx:w-1],%[fx:h-1],\#ff0 \
  -define quantum:format=floating-point -depth 32 \
  map.tiff
```

> I finally have a proper duvet cover now!

Sleep well!

**Reply by MarliesMaalderink** [reply]
> Sleep well!

Thanks!

In case anyone ever needs this -  eventually I ended up using photoshop for creating the warped displacement map anyway (one of my colleagues didn't have the "32 bit bug" so I could borrow her machine) because I couldn't get the right warping effect with Gimp. I really needed the photoshop warp tool. But in the end this worked fine too, as long as the initial tiff map is ok. As long as this map is openend in photoshop and nothing is changed about the file mode, it still works fine after saving.

Anyway, this is it (I kinda erased the edges very quickly, I haven't gotten to creating the proper overlay yet)

http://www.marliesmaalderink.nl/images/finalDuvetCover.jpg

**Reply by DOEHOONLEE** [reply]
Is there a way to run this on mac terminal?

```
magick.exe \
  -size 1000x1000 xc: \
  -sparse-color Bilinear \
0,0,\#000,\
%[fx:w-1],0,\#f00,\
0,%[fx:h-1],\#0f0,\
%[fx:w-1],%[fx:h-1],\#ff0 \
  -define quantum:format=floating-point -depth 32 \
  map.tiff
```

**Reply by fmw42** [reply]
@MarliesMaalderink  Can you elaborate on the Photoshop bug. I seem to be having an issue also getting 32 bit output from Photoshop to be correct using a HALD identity map after trying to warp it using Photoshop puppet warp

```
magick \
-size 1000x1000 xc: \
-sparse-color Bilinear \
0,0,"#008", \
"%[fx:w-1]",0,"#f08", \
0,"%[fx:h-1]","#0f8", \
"%[fx:w-1]","%[fx:h-1]","#ff8" \
-define quantum:format=floating-point -depth 32 \
identity_map.tiff
```

**Reply by MarliesMaalderink** [reply]
@fmw42  it is explained more thoroughly here: https://feedback.photoshop.com/conversations/photoshop/photoshop-cc-unable-to-edit-32-bitchannel-images/5f5f45654b561a3d42487caa?page=1

It ssems that sometimes photoshop can't work with 32-bit tiff files, I was unable to create new layers or use any sort of transform on it. My colleague had no problems with it on her computer. (we have a multi-node company subscription, all using the exact same version). I ended up installing it on my laptop and there it worked fine as well.

**Reply by fmw42** [reply]
This link does not work for me. Can you double check it?

**Reply by MarliesMaalderink** [reply]
@fmw42 It works for me. The title of the page is: "Photoshop CC: Unable to edit 32 bit/channel images", maybe you can access a version of it with google.

**Reply by DOEHOONLEE** [reply]
Hmm let me ask to be sure. So this `.tif` is recommended/mandatory for mapping on duvet/blanket? but does `.tif` have to be 32-bit in order to do so?

@MarliesMaalderink `access a version of it with google.` => so I can create a `.tif` using this?!

**Reply by DOEHOONLEE** [reply]
@fmw42 does imageMagick has some tutorial page on `.tiff` creating? On things like what it is supposed to contain, what it should look like, and etc. The `.tiff` that @MarliesMaalderink provided in this post is filled with colorwheel, but all other `.tiff` creating tutorial lectures I searched on YouTube seem like regular pictures. Maybe imageMagick has some sort of instructions on this?

Thanks!

**Reply by fmw42** [reply]
There is no tutorial that I know. It is an identity image for mapping as gradients, but using the color channels to hold the mapping rather than two separate images, one for x and one for y. A linear gradient means that the change in intensity is equal to the change in x or change in y. That means applying a gradient as a mapping leads to no change in the input image. So one starts with the gradients and applies some distortion to the gradients to create the mapping that will produce a change to the image when applied to it. See https://legacy.imagemagick.org/Usage/mapping/#displace_2d

**Reply by snibgo** [reply]
The format of the displacement/distortion map file (TIFF, MIFF, etc) doesn't matter. What does matter is the _precision_ of the data. 

I explore different aspects on various pages, such as:
[Displacement maps](http://im.snibgo.com/displace.htm)
[Follow line](http://im.snibgo.com/followline.htm)
[De-barrel distortion](http://im.snibgo.com/debarrel.htm)

**Comment by snibgo** [comment]
> finalDuvetCover.jpg

Good stuff, thanks for sharing that.

**Comment by DOEHOONLEE** [comment]
Great work @MarliesMaalderink !!

but.. following your steps, I got this with the following codes. I got same results from them, but unfortunately, neither of them work..

```
// [1]
magick '/Users/david/Documents/study/image-magick/src/ocean.png' -resize 1500x1500! '/Users/david/Documents/study/image-magick/src/testDisplace.jpeg' -compose Distort -composite '/Users/david/Documents/study/image-magick/src/result/FINALLY.png'

// [2]
convert '/Users/david/Documents/study/image-magick/src/blanket.jpg' \
'/Users/david/Documents/study/image-magick/src/testDisplace.jpeg' \
  -virtual-pixel transparent \
  -compose Distort -composite '/Users/david/Documents/study/image-magick/src/result/myBlanket.png'
```

![FINALLY](https://user-images.githubusercontent.com/20305442/131451072-b8e5c7d9-ae55-4692-8faa-1fd74f62239d.png)

but I cannot get rid of the greenish background and the ocean is not even fully mapped. Only part of the image is mapped inside the blanket shape.

Can you explain how you went from here?

Also, your final result seems much more realistic. Especially the circled part.

![sample](https://user-images.githubusercontent.com/20305442/131451520-d64a73ec-2327-40b8-b938-c9f34b83db0e.jpg)

Can you share how you got the wrinkled parts to work?

Thank you!!

**Reply by MarliesMaalderink** [reply]
There seem to be a few issues, the first being that you are using my original jpg as displacement map, which was the whole issue I was encountering, it should have been a 32-bit tiff.  Using a jpg causes the result to look fragmented. The jpg I uploaded was just to illustrate what I was trying to do, it has both the wrong size and it's not a 32-bit tiff.

The displacement and the source image need to be the same size. If they're not, the displacement will not match up and will use only part of the pixels of the source image.

About the wrinkles and the background, the final result has an overlay layer on top of it, hiding the background and creating the wrinkles. Overlays like this can be bought on several online shops, we bought this one on creativemarket.com (look for mockups)

**Reply by DOEHOONLEE** [reply]
In what way did you build your `.tiff` file? Is there some manual or instruction on creating `.tiff` for displacement mapping? Why is this in colorwheel? hmm..
